bentoml/_internal/frameworks/lightgbm.py

```diff
@@ -176,121 +166,96 @@ def save(
-    ):
-        super().__init__(tag, name=name)
-        self._booster_params = booster_params if booster_params is not None else {}
-        self._infer_api_callback = infer_api_callback
-
-        self._predict_fn: t.Any = None
-        self._model: t.Any = None
-
-    def _is_gpu(self):
-        try:
-            if "device" in self._booster_params:
-                assert isinstance(self._booster_params["device"], str)
-                return "gpu" in self._booster_params["device"]
-        except KeyError:
-            return False
-
-    @property
-    def num_replica(self) -> int:
-        if self._is_gpu() and self.resource_quota.nvidia_gpu:
-            return self.resource_quota.nvidia_gpu
-        return 1
-
-    def _setup(self) -> None:
-        self._model = load(
-            tag=self._tag,
-            booster_params=self._booster_params,
-            model_store=self.model_store,
-        )
-        self._predict_fn = getattr(self._model, self._infer_api_callback)
+        context=context,
+    ) as bento_model:
+        model.save_model(bento_model.path_of(MODEL_FILENAME))
 
-    def _run_batch(self, input_data: "ext.NpNDArray") -> "ext.NpNDArray":  # type: ignore[reportIncompatibleMethodOverride]
-        return self._predict_fn(input_data)
+        return bento_model.tag
 
 
-@inject
-def load_runner(
-    tag: t.Union[str, Tag],
-    infer_api_callback: str = "predict",
-    *,
-    booster_params: t.Optional[t.Dict[str, t.Union[str, int]]] = None,
-    name: t.Optional[str] = None,
-) -> "_LightGBMRunner":
+def get_runnable(bento_model: bentoml.Model) -> t.Type[bentoml.Runnable]:
+    """
+    Private API: use :obj:`~bentoml.Model.to_runnable` instead.
     """
-    Runner represents a unit of serving logic that can be scaled horizontally to
-    maximize throughput. :func:`bentoml.lightgbm.load_runner` implements a Runner class that
-    wrap around a LightGBM model, which optimize it for the BentoML runtime.
-
-    Args:
-        tag (:code:`Union[str, Tag]`):
-            Tag of a saved model in BentoML modelstore.
-        infer_api_callback (:code:`str`, `optional`, default to :code:`predict`):
-            Inference API callback from given model. If not specified, BentoML will use default :code:`predict`.
-            Users can also choose to use :code:`predict_proba` for supported model.
-        booster_params (:code:`Dict[str, Union[str, int]]`, `optional`, defaults to `None`):
-            Parameters for boosters. Refers to `Parameters Doc <https://lightgbm.readthedocs.io/en/latest/Parameters.html>`_
-            for more information.
-
-    Returns:
-        :obj:`~bentoml._internal.runner.Runner`: Runner instances for :mod:`bentoml.lightgbm` model
-
-    Examples:
 
-    .. code-block:: python
+    class LightGBMRunnable(bentoml.Runnable):
+        SUPPORT_NVIDIA_GPU = False  # afaik LightGBM only supports GPU during training, not for inference.
+        SUPPORT_CPU_MULTI_THREADING = True
```

Comment:
I don't believe LightGBM supports using a GPU for `.predict`, but please correct me if I'm mistaken!

Regarding cpu-multithreading support, I know LightGBM supports it, and I usually just have `n_jobs = -1` set in the model itself, which I believe utilizes all available cores. We could follow the XGBoost implementation and set `n_jobs` to the value of the envvar `OMP_NUM_THREADS`, but I don't know if that's really a common use case so I didn't implement it.

Reply:
You are correct, LightGBM doesn't have support for GPU, hence no need to support GPU here.

Boosting library such as LightGBM or XGBoost uses openmp to implements multi-threading, and hence we can control the behaviour via `OMP_NUM_THREADS`. I believe that we can mimic the behaviour from xgboost here for lightgbm, since these two are mostly similar.

Reply:
BentoML will set OMP_NUM_THREADS for each Runner process based on 1. the CPU cores available, 2. the runner resource configuration and 3. the cgroup settings in the context of Docker and Kubernetes. So in this case, I think you are right, we don't need to set the n_jobs or num_threads parameter, BentoML's Runner scheduling_strategy will take care of this by setting the `OMP_NUM_THREADS` env var (because SUPPORT_CPU_MULTI_THREADING is set to True), and it will be the optimal default configuration for lightGBM.